### ISSUE TITLE
1097 - IsUnion

### ISSUE THREAD

**Issue by bencor** [type-challenges/type-challenges#1140]
```ts
type IsUnionImpl<T, C extends T = T> = (T extends T ? C extends T ? true : unknown : never) extends true ? false : true;
type IsUnion<T> = IsUnionImpl<T>;
```

### Why does it work?

`type Distributive<T> = T extends T ? ConditionTrue<T> : never;` takes advantage of [Distributive Conditional Types](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types) to evaluate `ConditionTrue` separately for each value in a union type `T`. For example, `Distributive<string|number>` is evaluated as `ConditionTrue<string> | ConditionTrue<number>`. This works as `T extends MATCH ? ... : never` with any type `MATCH` for which the condition will always evaluate to true, such as `any` or `unknown`.

Let's re-write the solution with more details.
```ts
type DoubleDistribute<T, TRUE, FALSE, C = T> = T extends T ? C extends T ? TRUE : FALSE : never;
type IsUnionDetailed<T> = DoubleDistribute<T, true, unknown> extends true ? false : true;
```

`DoubleDistribute` has two distributive conditions: one for `T` and one for `C`, a copy of `T`. This effectively creates a nested loop where every type in the (potentially union) input type `T` is evaluated one by one against all others in `T` to return either `TRUE` or `FALSE`. The result of `DoubleDistribute` is then a union of all individual evaluations. Given that any type in `T` `extends` itself, a `TRUE` type is guaranteed in the resulting union. If `T` is a union type, each of its possible types will not `extends` all the others, leading to some `FALSE` types being included in the resulting union. Therefore, `DoubleDistribute` is either `TRUE` if `T` is a single (not union) type or `TRUE|FALSE` if `T` is a union. 

`IsUnionDetailed` looks at the result of `DoubleDistribute` as either `true` or `true|unknown` (which resolves to simply `unknown`) and returns the expected `true` or `false` return type.

### Example for single type:
```ts
IsUnion<string>
=> IsUnionImpl<string, string>
=> (string extends string ? string extends string ? true : unknown : never) extends true ? false : true
=> (string extends string ? true : unknown) extends true ? false : true
=> (true) extends true ? false : true
=> false
```

### Example for union type:
```ts
IsUnion<string|number>
=> IsUnionImpl<string|number, string|number>
=> (string|number extends string|number ? string|number extends string|number ? true : unknown : never) extends true ? false : true
=> (
  (string extends string|number ? string|number extends string ? true : unknown : never) |
  (number extends string|number ? string|number extends number ? true : unknown : never)
) extends true ? false : true
=> (
  (string|number extends string ? true : unknown) |
  (string|number extends number ? true : unknown)
) extends true ? false : true
=> (
  (
    (string extends string ? true : unknown) |
    (number extends string ? true : unknown)
  ) |
  (
    (string extends number ? true : unknown) |
    (number extends number ? true : unknown)
  )
) extends true ? false : true
=> (
  (
    (true) |
    (unknown)
  ) |
  (
    (unknown) |
    (true)
  )
) extends true ? false : true
=> (true|unknown) extends true ? false : true
=> (unknown) extends true ? false : true
=> true
```

**Comment by zhuo-huo-long** [comment]
Your article is wonderful and clear.

**Comment by doox911-opensource** [comment]
Test cases:

```ts
  Expect<Equal<IsUnion<(() => any)|(() => 15)>, true >>,
```
failed

**Comment by tenkirin** [comment]
@doox911-opensource 

I don't think `IsUnion<(() => any) | (() => 15)>` should equal to `true` as `() => any` is the widest type, a.k.a. the superset of function without any arguments, which certainly includes the `() => 15` in your case, and just like `any | 15` evaluates to `any`, `(() => any) | (() => 15)` evaluates to `() => any`, which is not a union.

**Comment by Lionad-Morotar** [comment]
Thank you !

**Comment by woai3c** [comment]
Hi, bencor. I don't know why `(true|unknown) extends true ? false : true` => `(unknown) extends true ? false : true`. 

Can you explain it? Thank you!

**Comment by wkz2003** [comment]
> 
unknown和其他类型一起做联合类型时会只留下unknown(除了和any一起)
unknown和其他类型一起做交叉类型时会得到去掉了unknown的类型
这有详细的
https://blog.csdn.net/qq_41632427/article/details/127060068

**Comment by wkz2003** [comment]
> Hi, bencor. I don't know why `(true|unknown) extends true ? false : true` => `(unknown) extends true ? false : true`.
> 
> Can you explain it? Thank you!

unknown和其他类型一起做联合类型时会只留下unknown(除了和any一起)
unknown和其他类型一起做交叉类型时会得到去掉了unknown的类型
这有详细的
https://blog.csdn.net/qq_41632427/article/details/127060068

**Comment by adoin** [comment]
汝甚吊，吾不如也

**Comment by diraneyya** [comment]
Excellent explanation.

```ts
type IsUnionDebug<T, TT = T> =
  T extends any ?
    TT extends T ?
      `'${TT}' extends '${T}'` : 
      `'${TT}' doesn't extends '${T}'` : never

type experiment = IsUnionDebug<'a' | 'b' | 'c'>

// This results in type experiment = "'a' doesn't extends 'b'" | "'a' doesn't extends 'c'" | 
// "'b' doesn't extends 'a'" | "'b' doesn't extends 'c'" | "'c' doesn't extends 'a'" | 
// "'c' doesn't extends 'b'" | "'a' extends 'a'" | "'b' extends 'b'" | "'c' extends 'c'"
```

**Comment by gearonix** [comment]
bro typescript is so damn hard

**Comment by nomoreyou** [comment]
> An attempt to decipher this great solution.
> 
> Ingredient 1: distributive conditional types. It is like iteration over the union and applying conditional to each element and combining the results back together as union.
> 
> ```ts
> type Foo<T> = T extends _some_condition_
> 
> Foo<A | B | C> 
> // translates to
> Foo<A> | Foo<B> | Foo<C>
> ```
> 
> Ingredient 2: rename type variable
> 
> ```ts
> T extends infer X ? ... : never // T is just renamed to X. `never` branch never hits
> ```
> 
> Ingredient 3: union never extends one element from that union. Only union of one element extends itself This is the crux of it. Checking if element of a union extend the union. If true than this union consist of one element.
> 
> ```ts
> A | B | C extends A // always false
> A extends A // always true. A is not a union. 
> ```
> 
> Putting all together. Take an element from a union and check if it extends this union.
> 
> ```ts
> type IsUnionImpl<T, C extends T = T> =
>   (T extends infer TItem // Iterate over T, here TItem is an item from the original union T. Ingredients 1&2
>     ? C extends TItem // C holds the original union T. Does union T extends an item from it? // Ingredient 3
>       ? true // yes. that could only be true if union T consist of one item
>       : unknown // no
>     : never) extends true ? false : true // have we got true from the above? yes - it's not a union
> 
> type IsUnion<T> = IsUnionImpl<T>
> ```

WOW  this is very clear 6666

**Comment by mbanem** [comment]
Looks like when T contains intersection of some elements then that intersection must be surrounded with elements with union, e.g. 
    <'a' | 'b' & 'c' & 'd' & 'e' | 'f'>
or to begin or end with at least two unions, 
    <'a' | 'b' | 'c' & 'd' & 'e' & 'f'> or
    <'a' & 'b' & 'c' & 'd' | 'e' | 'f'>

**Comment by heruiwoniou** [comment]
How about this ?

```ts
type IsUnion<T, K = T> = [T] extends [never]
  ? false
  : T extends K
  ? [K] extends [T]
    ? false
    : true
  : never;
```

**Comment by a1185963755** [comment]
1. **外层条件类型：** `(T extends T ? (C extends T ? true : unknown) : never) extends true`
   - `T extends T ? ... : ...` 是一个分布式条件类型，当 `T` 是联合类型时，这个条件会对联合类型的每个成员进行检查。

2. **内层条件类型：** `(C extends T ? true : unknown)`
   - `C` 是 `T` 的一个扩展类型，并且默认为 `T` 本身。因为 `C` 也是 `T` 的成员之一，所以这一步也会对联合类型的每个成员进行检查。



- 当 `T` 是一个联合类型（例如 `A | B`），分布式条件类型 `(T extends T ? ... : ...)` 会对联合类型的每个成员进行单独检查。
  - 假设 `T = A | B`，那么 `(A | B) extends (A | B)` 会被分解成 `A extends (A | B) ? (B extends (A | B) ? true : unknown) : never` 和 `B extends (A | B) ? (A extends (A | B) ? true : unknown) : never`。

- 如果 `T` 是一个联合类型，这些检查会产生不同的结果，因为 `C` 是联合类型的不同成员时的表现不同。
  - 对于联合类型的每个成员，检查 `C extends T ? true : unknown` 会产生 `true` 或 `unknown`。
  - 最终结果 `(true | unknown)` 不能扩展 `true`。

- 但是，如果 `T` 不是联合类型（例如是单个类型），那么 `T` 只有一个成员，`T extends T` 将总是 `true`，并且 `C extends T` 也总是 `true`，结果为 `true`。

- 因此，如果 `T` 是联合类型，上述条件的结果将不会是 `true`，而是某种形式的 `unknown`。但是，如果 `T` 不是联合类型，结果将是 `true`。

最终，这个类型的判断逻辑如下：
- 如果结果为 `true`，表示 `T` 不是联合类型。
- 如果结果不是 `true`，表示 `T` 是联合类型。

**Comment by Jayce-liang** [comment]
niu b plus

**Comment by holger2138** [comment]
按理讲 ` IsUnion<boolean | true>, IsUnion<string | ''>`  都就应该是 `true`, 但实际上是 `true, false`,有大佬解释下吗？

**Comment by heecheon92** [comment]
> Excellent explanation.
> 
> ```ts
> type IsUnionDebug<T, TT = T> =
>   T extends any ?
>     TT extends T ?
>       `'${TT}' extends '${T}'` : 
>       `'${TT}' doesn't extends '${T}'` : never
> 
> type experiment = IsUnionDebug<'a' | 'b' | 'c'>
> 
> // This results in type experiment = "'a' doesn't extends 'b'" | "'a' doesn't extends 'c'" | 
> // "'b' doesn't extends 'a'" | "'b' doesn't extends 'c'" | "'c' doesn't extends 'a'" | 
> // "'c' doesn't extends 'b'" | "'a' extends 'a'" | "'b' extends 'b'" | "'c' extends 'c'"
> ```

Such a brilliant way to debug a type....

**Comment by j-standfast** [comment]
This solution leverages distributive conditional types, as in the other solutions in this thread, without using a second type parameter.
```
type IsUnion<T> = 
  ([T] extends [never] ? never : T[]) extends (T extends any ? T[] : never) ? 
  false : 
  true;
```

more verbose / easier to understand / identical logic:
```
type DistributiveArray<T> = T extends any ? T[] : never;
type NondistributiveArray<T> = [T] extends [never] ? never : T[];
type IsUnion<T> = NondistributiveArray<T> extends DistributiveArray<T> ? false : true;
```

**Comment by baixiaoyu2997** [comment]
最简单的解释：

假设传入的联合类型是`'a'|'b'`
1. 当`T extends T`时，那么就是执行了`"a" extends "a"|"b"`和` "b" extends "a"|"b"` 
2. 再进行`C` extends `T`时，这个T就不再是`"a"|"b"`了，而是上一步通过的`"a"`或者`"b"`

如果说的不正确欢迎指正

**Comment by lumirelle** [comment]
```ts
type IsUnion<T, C extends T = T> = 
  [T] extends [never] 
    ? false
    : T extends T 
      ? Equal<C, T> extends true 
        ? false 
        : true 
      : true
```

**Comment by ffgenius** [comment]
> How about this ?
> 
> type IsUnion<T, K = T> = [T] extends [never]
>   ? false
>   : K extends T
>     ? [T] extends [K]
>       ? false
>       : true
>   : never;
> But I dont know why it does't work for `Equal<IsUnion<(() => any) | (() => 15)>, true>` 🤣

This classic `IsUnion` implementation hinges on `[B] extends [T]`.
In essence, it attempts to infer whether `B` is a union via assignability.

It relies on an implicit assumption:

> If `B` is a union, then for any distributed branch `T`, `B` should not be fully assignable to `T`.

This assumption holds in many common cases, for example:

```ts
[string | number] extends [string] // false
```

However, it is **not universally valid**.

When a union contains an *absorbing branch*, this check breaks down, for example:

- `(() => string) | (() => any)`
- `string | any`
- `boolean | true`

In these cases, even though `B` is clearly a union, there still exists some branch `T` such that:

```ts
[B] extends [T] // true
```

The issue is not with unions themselves, but with what `[B] extends [T]` actually tests.
It checks **whether the entire type is assignable to a single branch, while
assignability and “being a parallel union structure” are not equivalent concepts in TypeScript**.

As a result, this `IsUnion` approach only reliably handles *distinguishable unions without absorbing branches*.

When `any`, `unknown`, wider types in covariant positions (such as function return types), or special literals like `boolean` are involved, the algorithm reaches a theoretical undecidable boundary.

In other words, this is not an implementation bug, but a limitation of the TypeScript type system itself:
**it cannot express or observe “semantic unions” purely at the type level**.

Since I am not a native English speaker, this response has been translated and polished with the help of AI. I’m not sure whether it fully conveys my original intent.
